modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+  segment index within [0, here.maxTaskPar); those other segments likely are also
+  benefitted from local distribution from the round-robin algorithm discussed above.
+  We also steal a ratio of elements, say 25%, because it leaves all other segments
+  with 75% of their work; the more elements they have, the more we take, but the
+  less they have, the less we steal; this also has the added benefit of reducing
+  unnecessary work stealing between segments when the bag is nearly emptied. Furthermore,
+  stealing horizontally ensures that all segments remain filled and that we still
+  achieve parallelism across segments for removal operations. Lastly, we attempt
+  to steal a maximum of `N / sizeof(eltType)`, where N is some size in megabytes
+  (representing how much data can be sent in one network request), which keeps
+  down excessive communication.
+
+  This data structure does not come without flaws; as work stealing is dynamic
+  and triggered on demand, work stealing can still be performed in excess, which
+  dramatically causes a performance drop. Furthermore, a severe imbalance across
+  nodes, such as an unfair distribution of work to a single or small subset of nodes,
+  may also causes an equally severe performance degradation. To circumvent this,
+  we allow the user to determine manually when to 'balance', which will evenly distribute all
+  elements across nodes. This data structure scales in terms of nodes, processors per node,
+  and even work load. The larger the work load, the more data that gets stolen when
+  stealing work, and better locality of elements among segments. As well, to
+  achieve true parallelism, usage of a privatized instance is a requirement, as
+  it avoids the overhead of remotely accessing class fields, bounding performance
+  on communication.
+*/
+
+/*
+  Below are segment statuses, which is a way to make visible to outsiders the
+  current ongoing operation.
+*/
+private param STATUS_UNLOCKED : uint = 0;
+private param STATUS_ADD : uint = 1;
+private param STATUS_REMOVE : uint = 2;
+private param STATUS_LOOKUP : uint = 3;
+private param STATUS_BALANCE : uint = 4;
+
+/*
+  Below are statuses specific to the work stealing algorithm.
+*/
+private param WS_INITIALIZED = -1;
+private param WS_FINISHED_WITH_NO_WORK = 0;
+private param WS_FINISHED_WITH_WORK = 1;
+
+/*
+  The phases for operations.
+*/
+private param ADD_BEST_CASE = 0;
+private param ADD_AVERAGE_CASE = 1;
+private param REMOVE_BEST_CASE = 2;
+private param REMOVE_AVERAGE_CASE = 3;
+private param REMOVE_WORST_CASE = 4;
+
+/*
+  The initial amount of elements in an unroll block. Each successive unroll block
+  is double the size of it's predecessor, allowing for better locality for when
+  there are larger numbers of elements. The better the locality, the better raw
+  performance and easier it is to redistribute work.
+*/
+config param distributedBagInitialBlockSize = 1024;

```

Comment:
I remember originally it had to do with using tuples (which is no longer the case), I suppose that slipped my mind. I'll make that change.